gax-java/gax/src/main/java/com/google/api/gax/logging/Slf4jUtils.java

```diff
@@ -0,0 +1,269 @@
+      return LoggerFactory.getILoggerFactory();
+    }
+  }
+
+  // logging helper methods
+  static Map<String, Object> messageToMapWithGson(Message message)
+      throws InvalidProtocolBufferException {
+    String json = JsonFormat.printer().print(message);
+    return gson.fromJson(json, new TypeToken<Map<String, Object>>() {}.getType());
+  }
+
+  static void recordServiceRpcAndRequestHeaders(
+      String serviceName,
+      String rpcName,
+      String endpoint,
+      Map<String, String> requestHeaders,
+      LogData.Builder logDataBuilder,
+      LoggerProvider loggerProvider) {
+    LoggingUtils.executeWithTryCatch(
+        () -> {
+          Logger logger = loggerProvider.getLogger();
+          if (logger.isInfoEnabled()) {
+            addIfNotEmpty(logDataBuilder::serviceName, serviceName);
+            addIfNotEmpty(logDataBuilder::rpcName, rpcName);
+            addIfNotEmpty(logDataBuilder::httpUrl, endpoint);
+          }
+          if (logger.isDebugEnabled()) {
+            logDataBuilder.requestHeaders(requestHeaders);
+          }
+        });
+  }
+
+  private static void addIfNotEmpty(Consumer<String> setter, String value) {
+    if (value != null && !value.isEmpty()) {
+      setter.accept(value);
+    }
+  }
+
+  static void recordResponseHeaders(
+      Map<String, String> headers, LogData.Builder logDataBuilder, LoggerProvider loggerProvider) {
+    LoggingUtils.executeWithTryCatch(
+        () -> {
+          Logger logger = loggerProvider.getLogger();
+          if (logger.isDebugEnabled()) {
+            logDataBuilder.responseHeaders(headers);
+          }
+        });
+  }
+
+  static <RespT> void recordResponsePayload(
+      RespT message, LogData.Builder logDataBuilder, LoggerProvider loggerProvider) {
+    LoggingUtils.executeWithTryCatch(
+        () -> {
+          Logger logger = loggerProvider.getLogger();
+          if (logger.isDebugEnabled()) {
+            Map<String, Object> messageToMapWithGson =
+                Slf4jUtils.messageToMapWithGson((Message) message);
+
+            logDataBuilder.responsePayload(messageToMapWithGson);
+          }
+        });
+  }
+
+  static void logResponse(
+      String status, LogData.Builder logDataBuilder, LoggerProvider loggerProvider) {
+    LoggingUtils.executeWithTryCatch(
+        () -> {
+          Logger logger = loggerProvider.getLogger();
+          if (logger.isInfoEnabled()) {
+            logDataBuilder.responseStatus(status);
+          }
+          if (logger.isInfoEnabled() && !logger.isDebugEnabled()) {
+            Map<String, Object> responseData = logDataBuilder.build().toMapResponse();
+            Slf4jUtils.log(logger, Level.INFO, responseData, "Received Grpc response");
+          }
+          if (logger.isDebugEnabled()) {
+            Map<String, Object> responsedDetailsMap = logDataBuilder.build().toMapResponse();
+            Slf4jUtils.log(logger, Level.DEBUG, responsedDetailsMap, "Received Grpc response");
+          }
```

Comment:
I can't see to find the difference between the two if blocks besides the log level. Is this the only difference?

Reply:
Yes, you are right. the `responseData` content passed in is different because of previous `record..` steps, but in this step, only difference is log level.  Reason to put map creation step inside if block, is to get it guarded so it does not create if info/debug is not enabled.